packages/node-patches/src/fs.ts

```diff
@@ -0,0 +1,345 @@
+          }
+        };
+        origOpendir(...args);
+      } else {
+        return origOpendir(...args).then((dir: Dir) => {
+          return handleDir(dir);
+        });
+      }
+    };
+  }
+
+  async function handleDir(dir: Dir) {
+    const p = path.resolve(dir.path);
+    const origIterator = dir[Symbol.asyncIterator].bind(dir);
+    //tslint:disable-next-line:no-any
+    const origRead: any = dir.read.bind(dir);
+
+    dir[Symbol.asyncIterator] = async function*() {
+      for await (const entry of origIterator()) {
+        await handleDirent(p, entry);
+        yield entry;
+      }
+    };
+
+    //tslint:disable-next-line:no-any
+    (dir.read as any) = async (...args: any[]) => {
+      if (typeof args[args.length - 1] === 'function') {
+        const cb = args[args.length - 1];
+        args[args.length - 1] = async (err: Error, entry: Dirent) => {
+          cb(err, entry ? await handleDirent(p, entry) : null);
+        };
+        origRead(...args);
+      } else {
+        const entry = await origRead(...args);
+        if (entry) {
+          handleDirent(p, entry);
+        }
+        return entry;
+      }
+    };
+    //tslint:disable-next-line:no-any
+    const origReadSync: any = dir.readSync.bind(dir);
+    //tslint:disable-next-line:no-any
+    (dir.readSync as any) = () => {
+      return handleDirentSync(p, origReadSync());
+    };
+
+    return dir;
+  }
+
+  function handleDirent(p: string, v: Dirent): Promise<Dirent> {
+    return new Promise((resolve, reject) => {
+      if (!v.isSymbolicLink()) {
+        return resolve(v);
+      }
+      origReadlink(path.join(p, v.name), (err: Error, target: string) => {
+        if (err) {
+          return reject(err);
+        }
+
+        if (!isOutPath(path.resolve(target))) {
+          return resolve(v);
+        }
+
+        fs.stat(target, (err: Error & { code: string }, stat: Stats) => {
+          if (err) {
+            if (err.code === 'ENOENT') {
+              // this is a broken symlink
+              // even though this broken symlink points outside of the root
+              // we'll return it.
+              // the alternative choice here is to omit it from the directory listing altogether
+              // this would add complexity because readdir output would be different than readdir withFileTypes
+              // unless readdir was changed to match. if readdir was changed to match it's performance would be
+              // greatly impacted because we would always have to use the withFileTypes version which is slower.
+              return resolve(v);
+            }
+            // transient fs related error. busy etc.
+            return reject(err);
+          }
+
+          // add all stat is methods to Dirent instances with their result.
+          patchDirent(v, stat);
+          v.isSymbolicLink = () => false;
+          resolve(v);
+        });
+      });
+    });
+  }
+
+  function handleDirentSync(p: string, v: Dirent | null) {
```

Comment:
Would it be easier to get a patch the prototype, e.g. by getting an instance eagerly at the beginning?

Reply:
I haven't come up with any strategy that is actually simpler.

- Dir never passes the directory the Dirent is from to the Dirent which i need to make a correct decision in isSymbolicLink
- If i was to change Dir to set path thats also the same place i need to run realpath either sync or async so this logic wouldn't look different.

if Dir in node core has a single override-able method that created Dirents with arguments like syc/async and  the directory path that would be ideal. But because Dir's are instantiated in properties backed by private symbols i have to patch all 3 of the public methods.

the only change i need to make on  Dirent is to set a boolean to the retuen value of isSymbolicLink depending on the directory and the realpath.

Reply:
Ah, yes. I remember you talking about that now! Makes sense.